dbms/src/Flash/Coprocessor/DAGStorageInterpreter.cpp

```diff
@@ -485,6 +596,39 @@ void DAGStorageInterpreter::buildRemoteStreams(const std::vector<RemoteRequest>
     }
 }
 
+void DAGStorageInterpreter::buildRemoteSourceOps(
+    SourceOps & source_ops,
+    PipelineExecutorStatus & exec_status,
+    const std::vector<RemoteRequest> & remote_requests)
+{
+    std::vector<pingcap::coprocessor::CopTask> all_tasks = buildCopTasks(remote_requests);
+    const DAGSchema & schema = remote_requests[0].schema;
+    pingcap::kv::Cluster * cluster = tmt.getKVCluster();
+    bool has_enforce_encode_type = remote_requests[0].dag_request.has_force_encode_type() && remote_requests[0].dag_request.force_encode_type();
+    size_t concurrent_num = std::min<size_t>(context.getSettingsRef().max_threads, all_tasks.size());
+    size_t task_per_thread = all_tasks.size() / concurrent_num;
+    size_t rest_task = all_tasks.size() % concurrent_num;
+    pingcap::kv::LabelFilter tiflash_label_filter = pingcap::kv::labelFilterNoTiFlashWriteNode;
+    /// TODO: support reading data from write nodes
+    for (size_t i = 0, task_start = 0; i < concurrent_num; ++i)
+    {
+        size_t task_end = task_start + task_per_thread;
+        if (i < rest_task)
+            ++task_end;
+        if (task_end == task_start)
+            continue;
+        std::vector<pingcap::coprocessor::CopTask> tasks(all_tasks.begin() + task_start, all_tasks.begin() + task_end);
+
+        auto coprocessor_reader = std::make_shared<CoprocessorReader>(schema, cluster, tasks, has_enforce_encode_type, 1, tiflash_label_filter);
+        context.getDAGContext()->addCoprocessorReader(coprocessor_reader);
```

Comment:
Looks like a potential data race,
But it has no effect at present, because one mpptask will only have one table scan.

Reply:
My mistake... `context.getDAGContext()->addCoprocessorReader(coprocessor_reader);` must be done at compile time, because mpptask will use `CoprocessorReader` to do cancel.